modules/src/erc20.rs

```diff
@@ -0,0 +1,301 @@
+use odra::{
+    contract_env,
+    types::{event::OdraEvent, Address, Balance},
+    Mapping, Variable
+};
+
+use self::{
+    errors::Error,
+    events::{Approval, Transfer}
+};
+
+#[odra::module]
+pub struct Erc20 {
+    decimals: Variable<u8>,
+    symbol: Variable<String>,
+    name: Variable<String>,
+    total_supply: Variable<Balance>,
+    balances: Mapping<Address, Balance>,
+    allowances: Mapping<(Address, Address), Balance>
+}
+
+#[odra::module]
+impl Erc20 {
+    #[odra(init)]
+    pub fn init_with_supply(
+        &mut self,
+        symbol: String,
+        name: String,
+        decimals: u8,
+        initial_supply: Balance
+    ) {
+        let caller = contract_env::caller();
+
+        self.symbol.set(symbol);
+        self.name.set(name);
+        self.decimals.set(decimals);
+        self.total_supply.set(initial_supply);
+        self.balances.set(&caller, initial_supply);
+
+        if initial_supply > Balance::zero() {
+            Transfer {
+                from: None,
+                to: Some(caller),
+                amount: initial_supply
+            }
+            .emit();
+        }
+    }
+
+    #[odra(init)]
+    pub fn init(&mut self, symbol: String, name: String, decimals: u8) {
+        self.init_with_supply(symbol, name, decimals, Balance::zero());
+    }
+
+    pub fn transfer(&mut self, recipient: Address, amount: Balance) {
+        let caller = contract_env::caller();
+        self.raw_transfer(caller, recipient, amount);
+    }
+
+    pub fn transfer_from(&mut self, owner: Address, recipient: Address, amount: Balance) {
+        let spender = contract_env::caller();
+        self.spend_allowance(owner, spender, amount);
+
+        self.raw_transfer(owner, recipient, amount);
+    }
+
+    pub fn approve(&mut self, spender: Address, amount: Balance) {
+        let owner = contract_env::caller();
+
+        self.allowances.set(&(owner, spender), amount);
+        Approval {
+            owner,
+            spender,
+            value: amount
+        }
+        .emit();
+    }
+
+    pub fn name(&self) -> String {
+        self.name.get_or_default()
```

Comment:
```suggestion
        self.name.unwrap_or_revert_with(Error::NameNotSet)
```
would be more strict in case of module not being initialized.